### ISSUE TITLE
Add option to requireTrailingComma to ignore when closing brackets on the same line

### ISSUE THREAD

**Issue by panaggio** [jscs-dev/node-jscs#1302]
This is desired when using `requireTrailingComma`:

    foo({
        bar: 'baz',
        qux: 'quux',
    });

But I don't think this is:

    foo([{
        bar: 'baz',
    }, {
        qux: 'quux',
    }, ]);

I'd expect some option to allow me to write something like this:

    foo([{
        bar: 'baz',
    }, {
        qux: 'quux',
    }]);

**Comment by markelog** [comment]
That is weird, but we will accept PR for it

**Comment by mdevils** [comment]
I don't exactly understand what is the problem you are facing? Those two examples are just two different structures.

**Comment by panaggio** [comment]
@mdevils could you elaborate? I don't see the difference in structure.

**Comment by mdevils** [comment]
@panaggio Do you mean that in case of one property you don't want to have trailing comma?

**Comment by mdevils** [comment]
Oh, now I see the difference.
Looks like a very strange exception to me.

**Comment by panaggio** [comment]
@mdevils again, could you elaborate? What's weird? I want to be sure this isn't some bizarre nre rule value request before proposing any PR. If it's just me, I can patch my jscs and live with the patched version.

**Comment by hzoo** [comment]
Maybe the PR would help clarify?

I think its because the first example is different? Since people usually requireTrailingComma to help with diff reading... I think it's that the comma would seem unnecessary like `}, ]);`

```js
foo([{
    bar: 'baz',
}, {
    qux: 'quux',
}, ]); // the comma isn't the last token on that line

foo({
    bar: 'baz',
    qux: 'quux', // comma is the last token
});

// or you do this
foo([
{
    bar: 'baz',
},
{
    qux: 'quux',
},
]);
```

**Comment by panaggio** [comment]
> Maybe the PR would help clarify?

Maybe. I'll try to send something soon. Before that, even I need to understand if what I'm proposing makes sense or not. May this doesn't make any sense to anyone because my request really doesn't make any sense :(

I'm sure I need to clarify what I want. Sorry for all the confusing examples until now.

Let me try again.

> people usually requireTrailingComma to help with diff reading... 

That's true, and this is close enough to my desire/need, if that helps.

When using the proposed new (say) `allowThisSpecialCase` on `requireTrailingComma`, these should be accepted:

    var x = [{
        bar: 'baz',
    }, {
        qux: 'quux',
    }];

    var y = [[
        'bar',
        'baz',
    ], [
        'qux',
        'quux',
    ]];

Because it isn't. Why? A comma is required right before the last square bracket.

What I'd like to do is to be able to allow the last element on an array to be immediately followed by the closing square bracket without raising anything when using `requireTrailingComma`, because forcing the comma to be there on this coding style won't help with anything.

If I use `requireTrailingComma`, I'm now forced to chance may coding style to

    var x = [
        {
            bar: 'baz',
        }, {
            qux: 'quux',
        },
    ];

    var y = [
        [
            'bar',
            'baz',
        ], [
            'qux',
            'quux',
        ],
    ];

or to make it ugly like

    var x = [{
        bar: 'baz',
    }, {
        qux: 'quux',
    }, ];


    var y = [[
        'bar',
        'baz',
    ], [
        'qux',
        'quux',
    ], ];

Did this new explanation help in any way? Is this really some corner case more people could make use of?

**Comment by tp** [comment]
I agree with @panaggio. Having the same issue here where the format is desired in object literals, but not arrays.

Though maybe the formatting @hzoo presented should be used anyway.

**Comment by hzoo** [comment]
We released [3.0](https://github.com/jscs-dev/node-jscs/releases/tag/v3.0.0) recently! We also wrote a [post](https://medium.com/@markelog/jscs-end-of-the-line-bc9bf0b3fdb2#.oot3wvgpu) about how 3.0 will be our last major release (since we are deprecating the project) and that we are [merging with ESLint](http://eslint.org/blog/2016/04/welcoming-jscs-to-eslint)!

Thus we won't be working on any new features moving forward which includes new rules/options. We will continue to fix bugs and work on CST/better autofixing for current rules in the upcoming months both to help JSCS users and to gain experience that we can hopefully use in ESLint. But otherwise we both teams will be working on getting [ESLint feature parity with JSCS](https://github.com/eslint/eslint/milestones/JSCS%20Compatibility) and just a nice transition for our users.

I'm putting an `orphaned` label to note this issue isn't planned to be worked on, but might be a good contribution as a custom rule (not in JSCS core itself).

I would encourage anyone who wants the rule to try writing it themselves: I would either write a custom rule or submit an issue to ESLint and if accepted a PR. Otherwise a custom rule that won't be in core repo itself.

JSCS rules are also very similar to ESLint rules: here are some resources to get started: 

- [Project-specific Lint Rules with ESLint](https://gist.github.com/jareware/7179093)
- [Creating Custom Rules for ESLint](https://insideops.wordpress.com/2015/12/08/creating-custom-rules-for-eslint)
- [Maintaining JavaScript Code Quality with ESLint ](https://www.paypal-engineering.com/2014/12/12/maintaining-javascript-code-quality-with-eslint/)
